File: cpp/src/day02.cpp
// Copyright 2021 Stefan Boere <s.boere@hotmail.com>
#include "include/day02.h"
#include <fstream>
#include <iterator>
#include <tuple>
#include <vector>

enum Direction { Forward = 1, Up = 2, Down = 3 };

int ex02a_work(std::vector<std::tuple<Direction, int>> xs) {
  int depth = 0;
  int hor = 0;

  for (auto &el : xs) {
    auto dir = std::get<0>(el);
    auto x = std::get<1>(el);
    switch (dir) {
    case Direction::Forward:
      hor += x;
      break;
    case Direction::Up:
      depth -= x;
      break;
    case Direction::Down:
      depth += x;
      break;
    }
  }
  return depth * hor;
}

int ex02b_work(std::vector<std::tuple<Direction, int>> xs) {
  int depth = 0;
  int hor = 0;
  int aim = 0;

  for (auto &el : xs) {
    auto dir = std::get<0>(el);
    auto x = std::get<1>(el);

    switch (dir) {
    case Direction::Forward:
      depth += x * aim;
      hor += x;
      break;
    case Direction::Up:
      aim -= x;
      break;
    case Direction::Down:
      aim += x;
      break;
    }
  }
  return depth * hor;
}

std::tuple<int, int> ex02() {
  std::ifstream ex02_file("../input/02");
  std::vector<std::tuple<Direction, int>> ex02_input;

  std::string line;
  while (std::getline(ex02_file, line)) {
    size_t pos = line.find(' ');
    auto dirstr = line.substr(0, pos);
    auto numstr = line.substr(pos + 1);
    int num = std::stoi(numstr);

    Direction dir;
    if (dirstr == "forward")
      dir = Direction::Forward;
    else if (dirstr == "up")
      dir = Direction::Up;
    else if (dirstr == "down")
      dir = Direction::Down;

    std::tuple<Direction, int> comb(dir, num);
    ex02_input.push_back(comb);
  }
  ex02_file.close();

  int ex02a = ex02a_work(ex02_input);
  int ex02b = ex02b_work(ex02_input);

  std::tuple<int, int> result(ex02a, ex02b);

  return result;
}

File: rust/src/day02.rs
use std::fs;

enum Direction {
    Forward,
    Up,
    Down
}

pub fn ex02() -> (i32, i32) {
    let contents = fs::read_to_string("../input/02")
            .expect("File not found");

    let lns: Vec<(Direction, i32)> = contents.lines().map(parseln).collect();

    let a = ex02a_work(&lns);
    let b = ex02b_work(&lns);
    (a, b)
}

fn parseln(ln: &str) -> (Direction, i32) {
    let (xstr, ystr) = ln.split_once(' ').unwrap();
    let x = match xstr {
        "forward" => Some(Direction::Forward),
        "up" => Some(Direction::Up),
        "down" => Some(Direction::Down),
        _ => None
    }.unwrap();

    let y = ystr.parse().unwrap();
    (x, y)
}

fn ex02a_work(xs: &[(Direction, i32)]) -> i32 {

    let mut depth = 0;
    let mut hor = 0;
    for (dir, x) in xs {
        match dir {
          Direction::Forward => hor = hor + x,
          Direction::Up     => depth = depth - x,
          Direction::Down   => depth = depth + x
        }
    }

    depth * hor
}


fn ex02b_work(xs: &[(Direction, i32)]) -> i32 {

    let mut depth = 0;
    let mut hor = 0;
    let mut aim = 0;
    for (dir, x) in xs {
        match dir {
          Direction::Forward => {
              depth = depth + x * aim;
              hor = hor + x
          },
          Direction::Up     => aim = aim - x,
          Direction::Down   => aim = aim + x
        }
    }

    depth * hor
}


#[cfg(test)]
mod tests {
    use super::*;
    static EX02_EXAMPLE: [(Direction,i32);6] = [
        (Direction::Forward, 5),
        (Direction::Down, 5),
        (Direction::Forward, 8),
        (Direction::Up, 3),
        (Direction::Down, 8),
        (Direction::Forward, 2)
    ];

    #[test]
    fn ex02a_example() {
        let result = ex02a_work(&EX02_EXAMPLE);
        assert_eq!(result, 150);
    }

    #[test]
    fn ex02b_example() {
        let result = ex02b_work(&EX02_EXAMPLE);
        assert_eq!(result, 900);
    }

    #[test]
    fn ex02test() {
        let result = ex02();
        assert_eq!(result, (1815044, 1739283308));
    }
}
